### PR TITLE
Feature/documentation

### PR DESCRIPTION
**Pull request by andywong36**
Demo notebook for foundry!

This is a first pass at #16 - will update README.md with feedback from these notebooks. 

Shows some examples for 
1. Preprocessing
2. GLM's
3. Marginal Effects

image
image
image

### CODE REVIEW

File: docs/data/uci.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+from io import BytesIO
+from typing import List, Tuple
+from tempfile import TemporaryDirectory
+import requests
+import zipfile
+
+import pandas as pd
+
+def get_census_dataset() -> Tuple[pd.DataFrame, pd.Series]:
```
Comment (jwdink) [inline docs/data/uci.py:9]: can we update these functions to download locally and re-use that local download? here's an example from sklearn: https://github.com/scikit-learn/scikit-learn/blob/e58f36c73/sklearn/datasets/_twenty_newsgroups.py#L152

Comment (jwdink) [reply docs/data/uci.py:9]: then we could move into the package proper

File: foundry/evaluation/marginal_effects.py [hunk 2]
```diff
@@ -85,7 +85,18 @@ def feature_names_in(self) -> Sequence[str]:
...
+        relevant_features: List[List[str]] = (
+            list(
+                map(
+                    lambda str_or_list: [str_or_list] if isinstance(str_or_list, str) else list(str_or_list),
+                    col_transformer._columns
+                )
+            )
+        )
+        # flatten and drop duplicates
+        return list(set(sum(relevant_features, [])))
```
Comment (jwdink) [inline foundry/evaluation/marginal_effects.py:99]:
```suggestion
        feature_names_in = set()
        for cols in col_transformer._columns:
            if isinstance(cols, str):
                cols = [cols]
            for col in cols:
                feature_names_in.add(cols)
        return list(feature_names_in)
```

File: foundry/glm/glm.py [hunk 3]
```diff
@@ -54,6 +54,7 @@ class Glm(BaseEstimator):
      ``col_mapping={'loc':sklearn.compose.make_column_selector('^col+.'), 'scale':[col1]}``.
     :param sparse_mm_threshold: Density threshold for creating a sparse model-matrix. If X has density less than this,
      the model-matrix will be sparse; otherwise it will be dense. Default .05.
+     ``col_mapping={'loc':sklearn.compose.make_column_selector('^col.+'), 'scale':[col1]}``.
```
Comment (jwdink) [inline foundry/glm/glm.py:57]:
```suggestion
```

File: foundry/glm/glm.py [hunk 4]
```diff
@@ -247,10 +249,13 @@ def _fit(self,
...
              stopping: Union['Stopping', tuple, dict] = (.001,),
              max_iter: int = 200,
-             max_loss: float = 10.,
+             max_loss: float = np.inf,
              verbose: bool = True,
              estimate_laplace_coefs: bool = True) -> 'Glm':
 
+        if verbose:
+            print("Estimating model parameters")
+
```
Comment (jwdink) [inline foundry/glm/glm.py:258]: too verbose!

File: foundry/glm/glm.py [hunk 5]
```diff
@@ -318,22 +323,34 @@ def closure():
             except KeyboardInterrupt:
                 sleep(1)
                 break
+            except ValueError as e:
+                raise FitFailedException() from e
```
Comment (jwdink) [inline foundry/glm/glm.py:327]: can you explain this? I would think this is too broad

File: foundry/glm/glm.py [hunk 6]
```diff
@@ -318,22 +323,34 @@ def closure():
...
         if train_loss > max_loss:
             self._fit_failed += 1
-            raise FitFailedException(f"train_loss ({train_loss}) too high")
+            raise FitFailedException(f"train_loss ({train_loss}) too high, consider increasing max_loss ({max_loss})")
 
         self._fit_failed = 0
 
+        if verbose:
+            print("Estimating model parameters success!")
+
```
Comment (jwdink) [inline foundry/glm/glm.py:339]:
```suggestion
```

File: foundry/glm/glm.py [hunk 7]
```diff
@@ -318,22 +323,34 @@ def closure():
...
         if estimate_laplace_coefs:
             if verbose:
                 print("Estimating laplace coefs... (you can safely keyboard-interrupt to cancel)")
             try:
                 self.estimate_laplace_coefs(X=X, y=y, sample_weight=sample_weight)
             except KeyboardInterrupt:
-                pass
+                if verbose:
+                    print("Estimation laplace coefs interrupted")
```
Comment (jwdink) [inline foundry/glm/glm.py:347]:
```suggestion
```

File: foundry/glm/glm.py [hunk 8]
```diff
@@ -318,22 +323,34 @@ def closure():
...
+
+        if estimate_laplace_coefs and verbose:
+            if self.converged_:
+                print("Estimating laplace coefs success!")
+            else:
+                print("Estimating laplace coefs unsuccessful! See warnings / errors.")
```
Comment (jwdink) [inline foundry/glm/glm.py:353]:
```suggestion
```

File: foundry/glm/glm.py [hunk 9]
```diff
@@ -681,30 +703,34 @@ def estimate_laplace_coefs(self, X: ModelMatrix, y: ModelMatrix, sample_weight:
...
+            self.converged_ = False
+
             try:
-                cov = torch.inverse(hess)
-            except RuntimeError as e:
-                fail_msg = str(e)
+                # cov = torch.inverse(hess)  # Hello floating point errors my old friend.
+                cholesky_hess = torch.linalg.cholesky(hess)  # This is way more numerically stable
+                inv_cholesky_hess = torch.linalg.inv(cholesky_hess)
+                cov = inv_cholesky_hess.T @ inv_cholesky_hess
```
Comment (jwdink) [inline foundry/glm/glm.py:713]: just to clarify: you've confirmed that

```python
chol = torch.linalg.inv(torch.linalg.cholesky(hess))
cov = chol.T @ chol
```

 is more stable than
 
 ```python
 cov = torch.linalg.inv(hess)
 ```
 
 ? 
 
 I wonder why that is, since they both require inversion.
 
 If this is right, I think you can skip calculating `cov` and just pass `torch.distributions.MultivariateNormal(scale_tril=inv_cholesky_hess)`.

File: foundry/glm/glm.py [hunk 10]
```diff
@@ -681,30 +703,34 @@ def estimate_laplace_coefs(self, X: ModelMatrix, y: ModelMatrix, sample_weight:
...
+                self._coef_mvnorm_ = torch.distributions.MultivariateNormal(
+                    means, covariance_matrix=cov, validate_args=True
+                )
+
+                self.converged_ = True
+
+            except RuntimeError as e:
+                warn(f"RunTimeError(\"{(str(e))}\")")
+                warn("Second order estimation of parameter distribution failed. Constructing approximate covariance.")
+                fake_cov = torch.diag(torch.diag(hess).pow(-1).clip(min=1E-5))
```
Comment (jwdink) [inline foundry/glm/glm.py:733]: I have not so great feelings about constructing a fake-but-realistic covariance matrix. What would it be used for? Could this result in confusion?

File: foundry/util.py [hunk 11]
```diff
@@ -334,6 +335,8 @@ def to_2d(arr: ArrayType) -> ArrayType:
     :param arr: Numpy array or tensor.
     :return: 2d of same type
     """
+    # TODO what about pandas? Slicing arr[None, ...] not officially supported.
```
Comment (jwdink) [inline foundry/util.py:338]: Do we ever call this on a pandas object? The docs dont mention pandas support

**Review by jwdink — CHANGES_REQUESTED** [review]
Note: This is an initial review focusing on the code-changes and not yet on the documentation notebooks, which I need to give more thought to

File: foundry/glm/glm.py (hunk 10 above)
Comment (andywong36) [reply foundry/glm/glm.py:733]: In the same way that a hessian allows for second-order gradient descent, this fake-but-realistic is used for first-order gradient descent. 
Using the same logic, sampling from this is like a first-order estimation of the variation of model coefs.

File: docs/1 - Glm Intro.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+
+# -
+
+# ### Modelling heteroscedastic errors
+#
+# Above we can see how the Glm class can do linear regression. Ordinary least squares modelling typically assumes that the output variance is homoscedastic - i.e. that the variance $\sigma^2$ is constant, and that the error term $\epsilon = y - X\theta \sim N(0, \sigma^2)$. Sometimes that may not be the case, and we want to model the error as well. 
+#
+# We demonstrate this by adding some more noise to the outputs. This noise has zero mean, but variance that is a function of X. 
+
+# +
```
Comment (jwdink) [inline docs/1 - Glm Intro.py:102]: based on reading through these examples, it occurred to me we'd do well to have our own data-simulating function. I've added one on 64d5e224d098c13193047ecb9f0b350073ea58af

File: docs/2 - Foundry Intro.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+    InteractionFeatures(
+        [
+            (
+                make_column_selector("pass"), 
+                make_column_selector("onehot")
+            )
+        ]
+    ),
+    StandardScaler(),
+    Glm("categorical")
```
Comment (jwdink) [inline docs/2 - Foundry Intro.py:109]:
```suggestion
    Glm("bernoulli")
```

File: docs/2 - Foundry Intro.py [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+trained_model = classifier.fit(X, y, glm__estimate_laplace_coefs=False)
+
+# ## Evaluation - Marginal Effects
+
+from foundry.evaluation import MarginalEffects
+
+me = MarginalEffects(pipeline=trained_model, )
+me(
+    X, 
+    y.map(lambda item: 1 if item == ">50K" else 0).astype(float),
```
Comment (jwdink) [inline docs/2 - Foundry Intro.py:132]:
```suggestion
    (y == ">50K").astype(float), 
```

**Review by jwdink — COMMENTED** [review]
OK, took a closer look at the notebooks.

I think notebook-2 could be adapted to work well as a Quickstart guide. We'd probably want to add a bit more explanation/motivation, though. What are marginal-effects, and why are we using them here? I *think* the answer here would be something like "this is an easily explainable way to convey the many many coefficients being captured by the one-hot grade feature + interactions."

I think notebook-1 is probably a bit too in the weeds for a Quickstart guide, lmk if you disagree.

I'm going to keep thinking about good motivating examples. I'm trying to think of ones that answer the questions "why would you use foundry vs. base-sklearn or stats models?" To me, the answers to this are one or more of:

1) You're modeling a target with an unusual distribution
2) You want to perform statistical inference on your model (so no sklearn), but you also want the convenient/powerful sklearn API (so you can fit into the affordances of that API, e.g. `cross_val_*`)
3) You want to model interactions
4) You want a way to convey the effect/direction/contribution of individual features (MarginalEffects)

Of course, these ideas are all things I came up with *after* reading your PR and thinking about this a bit, so this isn't a ding on your notebooks. Thanks for your work here!

**Comment by jwdink** [comment]
Actual diffs since last merge can be found [here](https://github.com/strongio/foundry/pull/25/files/1a8ac78ccde094d26732407608a67868c7398149..f8067ca60ab286f9aac20e4944863bcdc7433e6f)

Summary of non-readme stuff:
- marginal-effects downsamples by default now (previously it did “marginal-at-medians”)
- marginal-effects uses geom_col for categorical features
- update NegativeBinomial parameterization so its first parameter is the mean
- Update Multinomial implementation so that torch’s limitations in `total_count` are only hit if you try to sample from the distribution
- move all the updated distributions into one module
- add `from_y` to Family, mostly so that we can stop passing `total_count` in X, which is weird (goodbye “asis_days_to_next_visit”?)
- fix bug in SliceDict (unserializing didn’t work)
- fix bug in ToSliceDict (“extras_ok” didn’t work)

**Comment by andywong36** [comment]
@jwdink all looks good for the README files! 🚀
